plugins/techdocs-addons/src/addons.tsx

```diff
@@ -0,0 +1,120 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  attachComponentData,
+  createReactExtension,
+  ElementCollection,
+  Extension,
+  useElementFilter,
+} from '@backstage/core-plugin-api';
+import React, { ComponentType, useCallback } from 'react';
+import { useOutlet } from 'react-router-dom';
+
+import { TechDocsAddonLocations, TechDocsAddonOptions } from './types';
+
+export const TECHDOCS_ADDONS_KEY = 'techdocs.addons.addon.v1';
+export const TECHDOCS_ADDONS_WRAPPER_KEY = 'techdocs.addons.wrapper.v1';
+
+/**
+ * TechDocs Addon registry.
+ * @public
+ */
+export const TechDocsAddons: React.ComponentType = () => null;
+
+attachComponentData(TechDocsAddons, TECHDOCS_ADDONS_WRAPPER_KEY, true);
+
+const getDataKeyByName = (name: string) => {
+  return `${TECHDOCS_ADDONS_KEY}.${name.toLocaleLowerCase('en-US')}`;
+};
+
+/**
+ * Create a TechDocs addon.
+ * @public
+ */
+export function createTechDocsAddon<TComponentProps>(
+  options: TechDocsAddonOptions<TComponentProps>,
+): Extension<ComponentType<TComponentProps>> {
+  const { name, component: TechDocsAddon } = options;
+  return createReactExtension({
+    name,
+    component: {
+      sync: (props: TComponentProps) => <TechDocsAddon {...props} />,
+    },
+    data: {
+      [TECHDOCS_ADDONS_KEY]: options,
+      [getDataKeyByName(name)]: true,
+    },
+  });
+}
+
+const getTechDocsAddonByName = (collection: ElementCollection, key: string) => {
+  return collection.selectByComponentData({ key }).getElements()[0];
+};
+
+const getAllTechDocsAddons = (collection: ElementCollection) => {
+  return collection
+    .selectByComponentData({
+      key: TECHDOCS_ADDONS_WRAPPER_KEY,
+    })
+    .selectByComponentData({
+      key: TECHDOCS_ADDONS_KEY,
+    });
+};
+
+const getAllTechDocsAddonsData = (collection: ElementCollection) => {
+  return collection
+    .selectByComponentData({
+      key: TECHDOCS_ADDONS_WRAPPER_KEY,
+    })
+    .findComponentData<TechDocsAddonOptions>({
+      key: TECHDOCS_ADDONS_KEY,
+    });
+};
+
+export const useTechDocsAddons = () => {
+  const node = useOutlet();
+
+  const collection = useElementFilter(node, getAllTechDocsAddons);
+  const options = useElementFilter(node, getAllTechDocsAddonsData);
+
+  const findAddonByData = useCallback(
+    (data: TechDocsAddonOptions | undefined) => {
+      if (!collection || !data) return null;
+      const nameKey = getDataKeyByName(data.name);
+      return getTechDocsAddonByName(collection, nameKey) ?? null;
+    },
+    [collection],
+  );
+
+  const renderComponentWithName = useCallback(
```

Comment:
sorry for nitpicking but now when I think about it maybe
renderComponentWithName --> renderComponentByName?